core/reflection.jvm/src/kotlin/reflect/jvm/internal/KCallableImpl.kt

```diff
@@ -112,65 +112,87 @@ internal abstract class KCallableImpl<out R> : KCallable<R>, KTypeParameterOwner
         return if (isAnnotationConstructor) callAnnotationConstructor(args) else callDefaultMethod(args, null)
     }
 
+    private val _absentArguments = ReflectProperties.lazySoft {
+        val parameterSize = parameters.size + (if (isSuspend) 1 else 0)
+        val maskSize = (parameters.size + Integer.SIZE - 1) / Integer.SIZE
+
+        // arguments with masks and DefaultConstructorMarker or MethodHandle
+        // +1 is argument for DefaultConstructorMarker or MethodHandle
+        val arguments = arrayOfNulls<Any?>(parameterSize + maskSize + 1)
+
+        // set absent values
+        parameters.forEach { parameter ->
+            if (parameter.isOptional && !parameter.type.isInlineClassType) {
+                // For inline class types, the javaType refers to the underlying type of the inline class,
+                // but we have to pass null in order to mark the argument as absent for InlineClassAwareCaller.
+                arguments[parameter.index] = defaultPrimitiveValue(parameter.type.javaType)
+            } else if (parameter.isVararg) {
+                arguments[parameter.index] = defaultEmptyArray(parameter.type)
+            }
+        }
+
+        for (i in 0 until maskSize) {
+            arguments[parameterSize + i] = 0
+        }
+
+        arguments
+    }
+
+    private fun getAbsentArguments(): Array<Any?> = _absentArguments().clone()
+
     // See ArgumentGenerator#generate
     internal fun callDefaultMethod(args: Map<KParameter, Any?>, continuationArgument: Continuation<*>?): R {
         val parameters = parameters
-        val arguments = ArrayList<Any?>(parameters.size)
-        var mask = 0
-        val masks = ArrayList<Int>(1)
-        var index = 0
-        var anyOptional = false
+        val parameterSize = parameters.size + (if (isSuspend) 1 else 0)
 
-        for (parameter in parameters) {
-            if (index != 0 && index % Integer.SIZE == 0) {
-                masks.add(mask)
-                mask = 0
+        val arguments = getAbsentArguments().apply {
+            if (isSuspend) {
+                // continuationArgument is tail of arguments
+                this[parameters.size] = continuationArgument
             }
+        }
+
+        // @see _parameters
+        val valueParameterOffset = when {
+            parameters.getOrNull(1)?.let { it.kind != KParameter.Kind.VALUE } ?: false -> 2
+            parameters.getOrNull(0)?.let { it.kind != KParameter.Kind.VALUE } ?: false -> 1
+            else -> 0
+        }
```

Comment:
The `Offset` is calculated first, and then subtracted from `parameter.index` only when the default argument is used, thereby reducing the amount of calculation.
[32cbe68](https://github.com/JetBrains/kotlin/pull/4840/commits/32cbe68ed38a0ff3f91f600d69dde94d1c0171c2)

Reply:
TBH this is more difficult to read, and also will probably have to be deleted anyway in the future when we support [context receivers](https://github.com/Kotlin/KEEP/issues/259), so I'd suggest to leave the old way.

Reply:
Reverted.
[1e1cd3a](https://github.com/JetBrains/kotlin/pull/4840/commits/1e1cd3a26a59bcd0ac3a1e50fc23873f16802851)